clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -237,12 +239,18 @@ public void close() {
      * @param send The request to send
      */
     public void send(Send send) {
-        KafkaChannel channel = channelOrFail(send.destination());
-        try {
-            channel.setSend(send);
-        } catch (CancelledKeyException e) {
-            this.failedSends.add(send.destination());
-            close(channel);
+        String connectionId = send.destination();
+        KafkaChannel closingChannel = closingChannels.remove(connectionId);
+        if (closingChannel != null)
+            doClose(closingChannel, true);
```

Comment:
Hmm, I wonder if we should change `private void close(KafkaChannel channel, boolean processOutstanding)` to take an additional `notifyDisconnect` parameter and then have an overload that defaults `notifyDisconnect` to be the same as `processOutstanding`. What do you think?